### PR TITLE
Clean up handling of max_page_size_bytes in Parquet writer

### PR DESCRIPTION
**Pull request by etseidl**
## Description
Reverts a change made in #12182 which exposed some potential issues with calculating page boundaries.  Instead, checks are added to ensure that page sizes will not exceed what can be represented by a signed 32-bit integer.  Also fixes some bounds checking that is no longer correct given changes made to the max page fragment size, and changes the column_index_truncation_length from `size_type` (which was used incorrectly) to `int32_t`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vuule** [comment]
ok to test

File: cpp/src/io/parquet/page_enc.cu [hunk 1]
```diff
@@ -200,6 +219,10 @@ __global__ void __launch_bounds__(block_size)
       len = block_reduce(reduce_storage).Sum(len);
       if (t == 0) { s->frag.fragment_data_size += len; }
       __syncthreads();
+      // page fragment size must fit in a 32-bit signed integer
+      if (s->frag.fragment_data_size > INT32_MAX) {
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:223]: IMO `numeric_limits` is preferred.
```suggestion
      if (s->frag.fragment_data_size > std::numeric_limits<int32_t>::max()) {
```

**Review by vuule — COMMENTED** [review]
Looks good, just one suggestion

File: cpp/src/io/parquet/page_enc.cu (hunk 1 above)
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu]: I'll hunt down the rest of them

**Review by vuule — APPROVED** [review]

**Comment by ttnghia** [comment]
>  changes the column_index_truncation_length from size_type (which was used incorrectly) to int32_t.

Can you clarify why `size_type` is incorrect, please?

**Review by bdice — APPROVED** [review]

**Comment by bdice** [comment]
> > changes the column_index_truncation_length from size_type (which was used incorrectly) to int32_t.
> 
> Can you clarify why `size_type` is incorrect, please?

I'm guessing it's because the use of `int32_t` here is defined by the Parquet format, and it is not dependent on cuDF's internal definition of `using size_type = int32_t;`. Is this correct?

**Comment by etseidl** [comment]
> I'm guessing it's because the use of `int32_t` here is defined by the Parquet format, and it is not dependent on cuDF's internal definition of `using size_type = int32_t;`. Is this correct?

Yep, you got it.  As I understand it, `size_type` is for indexing purposes inside libcudf.  The parameter in question should be an `int32_t` regardless of the definition of `size_type` (which could change in the future for all I know).

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**